### PR TITLE
#166272334 Verify User Account

### PR DESCRIPTION
**Pull request by emmygozi**
#### What does this PR do?

This pull request verifies a new user on signup or on login.
It caters for scenarios when verification link expires by creating two new APIs to generate verification link and verify a user respectively.
#### Description of Task to be completed?

- create API on sendgrid
- send verification mail on signup
- API to generate verification mail in case of link expiration
- create verify mail model and migration
- verifies a new user
- Sync with Macco to use mail generating template
- Write unit tests to improve test coverage and test uncovered files
#### Any background context you want to provide?

Had to work in tandem with Macco to agree on templates to use
#### What are the relevant pivotal tracker stories?

https://www.pivotaltracker.com/story/show/166272334
#### Screenshots (if appropriate)

image

#### Questions:

N/A

### CODE REVIEW

File: server/controllers/users.js [hunk 1]
```diff
@@ -37,7 +43,19 @@ class UserController {
...
+      const verifyToken = randomstring.generate(40);
+
+      const verifyDetails = {
+        verifyToken, tokenExpiry, userId: user.id
+      };
+
+      await VerifyUser.create({ ...verifyDetails });
+      sendVerifyMailToken(verifyToken, user.email, user.username);
+      return res.status(201).json({
+        status: 'success', message: 'User created successfully', user: userExtractor(user, token), verifyToken
```
Comment (fidelisojeah) [inline server/controllers/users.js]: why exactly are you sending the `verifyToken` ?

File: server/controllers/users.js [hunk 2]
```diff
@@ -152,6 +169,91 @@ class UserController {
...
+  * @param {object} res - Response object
+  * @param {object} next The next middleware
+  * @return {json} Returns json object
+  * @static
+  */
+  static async verifyAccount(req, res, next) {
+    try {
+      const { token, email } = req.query;
+
+      if (!token || !email) return Response.error(res, 400, 'Please use a valid reset link');
```
Comment (fidelisojeah) [inline server/controllers/users.js]: reset?

File: server/controllers/users.js [hunk 3]
```diff
@@ -152,6 +169,91 @@ class UserController {
...
+      const user = await User.findOne({ where: { email } });
+      if (!user) return Response.error(res, 404, 'User does not exist');
```
Comment (fidelisojeah) [inline server/controllers/users.js]: lol what?

firstly, this wouldn't exactly be a 404 (considering the username is not in the endpoint)
Also, you are releasing too much information by doing this.
`Invalid Token` would work here

File: server/helpers/mailer.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import sgMail from '@sendgrid/mail';
+import Mailgen from 'mailgen';
+import { config } from 'dotenv';
+
+config();
+
+const url = process.env.UI_HOST_NAME || 'https://kingsmen-ah-frontend-staging.herokuapp.com';
```
Comment (fidelisojeah) [inline server/helpers/mailer.js]: nah

you would be better off defaulting this to localhost.

**Review by fidelisojeah — CHANGES_REQUESTED** [review]
just minor changes

File: server/controllers/users.js (hunk 2 above)
Comment (emmygozi) [reply server/controllers/users.js]: Would fix this. Typo.

File: server/controllers/users.js (hunk 3 above)
Comment (emmygozi) [reply server/controllers/users.js]: Alright on it.

File: server/controllers/users.js (hunk 1 above)
Comment (emmygozi) [reply server/controllers/users.js]: For my test to have access to the verification token.

Comment (fidelisojeah) [reply server/controllers/users.js]: Remove it.

Find another way to write your tests

**Review by fidelisojeah — DISMISSED** [review]

**Review by adex001 — DISMISSED** [review]

**Review by GerrardE — APPROVED** [review]
Good One

**Review by adex001 — APPROVED** [review]

**Pull request merged**